File: python/MCDWrapper.py
import numpy as np
import cv2
import KLTWrapper
import ProbModel
from PIL import Image
import imagehash
from skimage.metrics import structural_similarity as ssim
def mse(image1, image2):
    # 计算均方误差
    # abs = (image1.astype("float") - image2.astype("float"))**2
    # abs = abs.flatten()
    # abs = abs[abs > 100]
    # err = abs.mean()
    err = (image1.astype("float") - image2.astype("float")) ** 2
    err = err.flatten()
    # err = err[err > 4]
    err = err.mean()
    # err = np.sum(err)
    # err /= float(image1.shape[0] * image1.shape[1])
    return err


def dhash(image, hash_size=4):
    image = Image.fromarray(image)
    return imagehash.average_hash(image, hash_size=hash_size)


def hamming_distance(hash1, hash2, hash_size=4):
    distance = hash1 - hash2
    similarity = 1 - (distance / (hash_size * hash_size))
    return similarity


def get_region(point, image_width, image_height):
    x, y = point
    region_width = image_width // 4
    region_height = image_height // 4
    row = y // region_height
    col = x // region_width
    return row * 4 + col


def calculate_intersection_area(rectangle1, rectangle2):
    x1, y1, w1, h1 = rectangle1
    x2, y2, w2, h2 = rectangle2
    intersection_x = max(0, min(x1 + w1, x2 + w2) - max(x1, x2))
    intersection_y = max(0, min(y1 + h1, y2 + h2) - max(y1, y2))
    intersection_area = intersection_x * intersection_y
    return intersection_area


def calculate_union_area(rectangle1, rectangle2):
    x1, y1, w1, h1 = rectangle1
    x2, y2, w2, h2 = rectangle2
    area1 = w1 * h1
    area2 = w2 * h2
    union_area = area1 + area2 - calculate_intersection_area(rectangle1, rectangle2)
    return union_area


def calculate_iou(bbox, track_box):
    intersection_area = calculate_intersection_area(bbox, track_box)
    union_area = calculate_union_area(bbox, track_box)
    iou = intersection_area / union_area
    return iou


def combine_rectangles(rectangles):
    combined_rectangles = []
    for i in range(len(rectangles)):
        is_overlapping = False
        rect = rectangles[i]
        for oth_rect in rectangles:
            if rect == oth_rect:
                continue
            combined_rect = [0, 0, 0, 0]
            combined_rect[0] = min(oth_rect[0], rect[0])
            combined_rect[1] = min(oth_rect[1], rect[1])
            x1 = max(oth_rect[0] + oth_rect[2], rect[0] + rect[2])
            y1 = max(oth_rect[1] + oth_rect[3], rect[1] + rect[3])
            if x1 - combined_rect[0] <= oth_rect[2] + rect[2] and y1 - combined_rect[1] <= oth_rect[3] + rect[3]:
                combined_rect[2] = x1 - combined_rect[0]
                combined_rect[3] = y1 - combined_rect[1]
                combined_rectangles.append(combined_rect)
                is_overlapping = True
                break
        if not is_overlapping:
            combined_rectangles.append(rect)
    combined_rectangles = np.unique(combined_rectangles, axis=0)
    return combined_rectangles


class tracker:
    def __init__(self, x, y, w, h):
        self.bgs_hitCounter = 0
        self.bgs_missCounter = 0
        self.hitCounter = 0
        self.missCounter = 0
        self.update_status = True
        self.x = x
        self.y = y
        self.w = w
        self.h = h
        self.box = (x, y, w, h)

    def calculate_iou(self, bbox):
        intersection_area = calculate_intersection_area(bbox, self.track_box)
        union_area = calculate_union_area(bbox, self.track_box)
        iou = intersection_area / union_area
        return iou

    def update(self, bbox):
        self.box = bbox
        self.update_status = True


class MCDWrapper:
    def __init__(self):
        self.imgIpl = None
        self.imgGray = None
        self.imgGrayPrev = None
        self.frm_cnt = 0
        self.lucasKanade = KLTWrapper.KLTWrapper()
        self.model = ProbModel.ProbModel()
        self.tracked_list = None
        self.bgs_tracked_list = None
        self.fream_num = None

    def init(self, image):
        self.imgGray = image
        self.imgGrayPrev = image
        self.lucasKanade.init(self.imgGray)
        self.model.init(self.imgGray)
        self.tracked_list = []
        self.bgs_tracked_list = []
        self.fream_num = 0

    def run(self, frame):
        framediff_res = []
        background_res = []
        bgs_tracked_res = []
        fd_tracked_res = []
        self.frm_cnt += 1
        self.imgIpl = frame
        self.imgGray = frame
        if self.imgGrayPrev is None:
            self.imgGrayPrev = self.imgGray.copy()

        self.lucasKanade.RunTrack(self.imgGray, self.imgGrayPrev)
        if self.fream_num == 119:
            self.model.alexinit(frame)

        self.model.motionCompensate(self.lucasKanade.H)

        if self.fream_num == 130:
            self.model.exchange_model()
            self.fream_num = 0

        bgs_mask = self.model.update(frame)

        kernel = np.ones((5, 5), np.uint8)
        thresh = cv2.dilate(bgs_mask, kernel, iterations=1)
        thresh = cv2.erode(thresh, kernel, iterations=1)
        contours, hierarchy = cv2.findContours(thresh.copy(), cv2.RETR_EXTERNAL, cv2.CHAIN_APPROX_SIMPLE)
        for contour in contours:
            (x, y, w, h) = cv2.boundingRect(contour)
            if cv2.contourArea(contour) < 40:
                continue
            background_res.append((x, y-2, w, h+2))
        for tracked in self.bgs_tracked_list:
            tracked.update_status = False
        if len(self.bgs_tracked_list) == 0:
            for contour in contours:
                if cv2.contourArea(contour) < 40:
                    continue
                (x, y, w, h) = cv2.boundingRect(contour)
                self.bgs_tracked_list.append(tracker(x, y-2, w, h+2))
        else:
            for box in background_res:
                num = get_region((box[0], box[1]), 1920, 1080)
                new_point = cv2.perspectiveTransform(np.asarray([box[0], box[1]]).reshape(-1, 1, 2).astype(np.float32),
                                                     self.lucasKanade.H[num]).reshape(-1)
                matched = False
                for tracked in self.bgs_tracked_list:
                    if calculate_iou((new_point[0],new_point[1],box[2],box[3]), tracked.box) > 0.4:
                # matched = False
                # for tracked in self.bgs_tracked_list:
                #     if calculate_iou(box, tracked.box) > 0.4:
                        tracked.update(box)
                        matched = True
                        tracked.hitCounter += 1
                        tracked.missCounter = 0
                if not matched:
                    self.bgs_tracked_list.append(tracker(box[0], box[1], box[2], box[3]))
            self.bgs_tracked_list = [obj for obj in self.bgs_tracked_list if obj.missCounter < 3]
            for i in range(len(self.bgs_tracked_list)):
                if not self.bgs_tracked_list[i].update_status:
                    self.bgs_tracked_list[i].missCounter += 1
                    # bgs_tracked_res.append(self.bgs_tracked_list[i].box)
                    continue
            # self.bgs_tracked_list = [obj for obj in self.bgs_tracked_list if obj.missCounter < 3]
            # for i in range(len(self.bgs_tracked_list)):
                if self.bgs_tracked_list[i].hitCounter > 2:
                    bgs_tracked_res.append(self.bgs_tracked_list[i].box)

        bgs_tracked_res = np.unique(bgs_tracked_res, axis=0)
        bgs_tracked_list_point = []
        new_bgs_tracked_res = []
        for i in bgs_tracked_res:
            num = get_region((i[0], i[1]), 1920, 1080)
            new_point = cv2.perspectiveTransform(np.asarray([i[0], i[1]]).reshape(-1, 1, 2).astype(np.float32),
                                                 self.lucasKanade.H[num]).reshape(-1)
            if new_point[0] < 0:
                new_point[0] = 0
            if new_point[0] >= 1920:
                new_point[0] = 1920 - 1
            if new_point[1] < 0:
                new_point[1] = 0
            if new_point[1] >= 1080:
                new_point[1] = 1080 - 1
            bgs_tracked_list_point.append([int(new_point[0]), int(new_point[1]), i[2], i[3]])

            old_bbox = self.imgGray[i[1]:i[1] + i[3], i[0]:i[0] + i[2]]
            new_bbox = self.imgGrayPrev[int(new_point[1]):int(new_point[1]) + i[3],
                       int(new_point[0]):int(new_point[0]) + i[2]]
            if old_bbox.shape[0] != new_bbox.shape[0] or old_bbox.shape[1] != new_bbox.shape[1]:
                continue
            else:
                # hash1=dhash(old_bbox)
                # hash2=dhash(new_bbox)
                # rr=hamming_distance(hash1, hash2)
                # rr=ssim(old_bbox, new_bbox)
                # print(rr)
                rr = mse(old_bbox, new_bbox)
                if rr > 250:
                    new_bgs_tracked_res.append(i)

        self.imgGrayPrev = self.imgGray.copy()
        self.fream_num += 1
        # return bgs_tracked_res,fd_tracked_res
        new_bgs_tracked_res = np.unique(new_bgs_tracked_res, axis=0)
        return background_res,bgs_tracked_res, new_bgs_tracked_res,thresh
        # return bgs_tracked_res, bgs_tracked_list_point,thresh

File: src/MCDWrapper.cpp
#ifndef _MCDWRAPPER_CPP_
#define _MCDWRAPPER_CPP_

#include "MCDWrapper.hpp"
#include "params.hpp"
#include <cstring>
#include <ctime>

MCDWrapper::MCDWrapper() {}

MCDWrapper::~MCDWrapper() { delete h; }

void MCDWrapper::Init(const UMat &imgGray) {

  frm_cnt = 0;
  Mat mat_imgGray;
  imgGray.copyTo(mat_imgGray);
  m_LucasKanade.Init(imgGray.cols, imgGray.rows);
  BGModel.init(mat_imgGray);
  imgGrayPrev = imgGray.clone();
  h = new double[16 * 9];
}
cv::Point2f MCDWrapper::compensate(cv::Point2f point, double *h) {
  float newW = 0;
  float newX = 0;
  float newY = 0;
  int modelWidth = imgGrayPrev.cols;
  int modelHeight = imgGrayPrev.rows;
  int X = point.x;
  int Y = point.y;

  int h_idxi = X / 480;
  int h_idxj = Y / 270;
  int h_id = h_idxi + h_idxj * 4;

  newW = h[h_id * 9 + 6] * X + h[h_id * 9 + 7] * Y + h[h_id * 9 + 8];
  newX = (h[h_id * 9 + 0] * X + h[h_id * 9 + 1] * Y + h[h_id * 9 + 2]) / newW;
  newY = (h[h_id * 9 + 3] * X + h[h_id * 9 + 4] * Y + h[h_id * 9 + 5]) / newW;
  return cv::Point2f(newX, newY);
}

std::vector<cv::Rect> MCDWrapper::Run(const UMat &imgGray) {
  /*************************
    计算KLT 单应矩阵H 模型补偿
  **************************/
  m_LucasKanade.RunTrack(imgGray, imgGrayPrev);
  m_LucasKanade.GetHomography(h);
  Mat Temp;
  imgGray.copyTo(Temp);
  BGModel.motionCompensate(h, Temp);
  Mat H(3, 3, CV_32F);
  memcpy(H.data, h, sizeof(float) * 9);

  /*************************
      模型更新 前景提取
  **************************/
  Mat Temp1;
  imgGray.copyTo(Temp1);
  BGModel.update(Temp1);
  /*************************
      形态学处理并绘制初始框
  **************************/
  vector<vector<Point>> contours;
  vector<Vec4i> hierarchy;
  vector<Rect> bgs_tracked_res;
  vector<Rect> bgs_tracked_list_point;
  vector<Rect> new_bgs_tracked_res;

  UMat thresh;
  thresh = UMat::zeros(imgGray.rows, imgGray.cols, CV_8UC1);
  Mat kernel = getStructuringElement(MORPH_RECT, Size(5, 5));
  dilate(BGModel.mask, thresh, kernel, Point(-1, -1), 1);
  erode(thresh, thresh, kernel, Point(-1, -1), 1);
  background_res.clear();
  findContours(thresh.clone(), contours, hierarchy, RETR_EXTERNAL,
               CHAIN_APPROX_SIMPLE);
  for (size_t i = 0; i < contours.size(); i++) {
    if (contourArea(contours[i]) < 40)
      continue;
    Rect rect = boundingRect(contours[i]);
    background_res.push_back(rect);
  }
  /************************************
      tracker追踪 通过运动连续性对运动目标框进行筛选
  *************************************/
  //更新所有trackerbox状态
  for (size_t i = 0; i < bgs_tracked_list.size(); i++) {
    bgs_tracked_list[i].update_status = false;
  }
  //初始化tracker list
  if (bgs_tracked_list.empty()) {
    for (size_t i = 0; i < contours.size(); i++) {
      if (contourArea(contours[i]) < 40)
        continue;
      Rect bbox = boundingRect(contours[i]);
      Tracker tracker(bbox.x, bbox.y - 2, bbox.width, bbox.height + 2);
      bgs_tracked_list.push_back(tracker);
    }
  } else {
    //更新tracker list中匹配框的坐标,将未匹配的加入list
    for (size_t i = 0; i < background_res.size(); i++) {
      Rect box = background_res[i];
      cv::Point2f new_point = compensate(cv::Point2f(box.x, box.y), h);
      bool matched = false;
      for (size_t j = 0; j < bgs_tracked_list.size(); j++) {
        if (bgs_tracked_list[j].calculate_iou(
                Rect(new_point.x, new_point.y, box.width, box.height)) > 0.4) {
          bgs_tracked_list[j].update(box);
          matched = true;
          bgs_tracked_list[j].hitCounter++;
          bgs_tracked_list[j].missCounter = 0;
        }
      }
      if (!matched) {
        Tracker tracker(box.x, box.y, box.width, box.height);
        bgs_tracked_list.push_back(tracker);
      }
    }

    //删除tracker list中missCounter >= 3的box
    vector<Tracker>::iterator iter;
    for (iter = bgs_tracked_list.begin(); iter != bgs_tracked_list.end();) {
      if (iter->missCounter >= 3) {
        iter = bgs_tracked_list.erase(iter);
      } else {
        iter++;
      }
    }

    //更新tracker list中box的missCounter,hitCounter,将结果加入bgs_tracked_res
    for (size_t i = 0; i < bgs_tracked_list.size(); i++) {
      if (!bgs_tracked_list[i].update_status) {
        bgs_tracked_list[i].missCounter++;
        continue;
      }
      if (bgs_tracked_list[i].hitCounter > 2) {
        bgs_tracked_res.push_back(bgs_tracked_list[i].box);
      }
    }
  }

  //计算unique删除重复的box
  unique_bgs_tracked_res.clear();
  for (size_t i = 0; i < bgs_tracked_res.size(); i++) {
    bool is_unique = true;
    for (size_t j = 0; j < unique_bgs_tracked_res.size(); j++) {
      if (bgs_tracked_res[i] == unique_bgs_tracked_res[j]) {
        is_unique = false;
        break;
      }
    }
    if (is_unique) {
      unique_bgs_tracked_res.push_back(bgs_tracked_res[i]);
    }
  }
  /************************************
      计算均方误差追踪 通过相似性对运动目标框进行筛选
  *************************************/
  new_bgs_tracked_res.clear();
  for (size_t i = 0; i < unique_bgs_tracked_res.size(); i++) {
    Rect box = unique_bgs_tracked_res[i];
    cv::Point2f new_point = compensate(cv::Point2f(box.x, box.y), h);
    float newX = new_point.x;
    float newY = new_point.y;
    if (newX < 0) {
      newX = 0;
    }
    if (newX + box.width >= imgGray.cols) {
      newX = imgGray.cols - box.width - 1;
    }
    if (newY < 0) {
      newY = 0;
    }
    if (newY + box.height >= imgGray.rows) {
      newY = imgGray.rows - box.height - 1;
    }
    if (box.height >= imgGray.rows || box.width >= imgGray.cols) {
      continue;
    }
    bgs_tracked_list_point.push_back(
        Rect(int(newX), int(newY), box.width, box.height));

    UMat old_bbox_tmp = imgGray(box);
    UMat new_bbox_tmp =
        imgGrayPrev(Rect(int(newX), int(newY), box.width, box.height));
    Mat old_bbox;
    Mat new_bbox;
    old_bbox_tmp.copyTo(old_bbox);
    new_bbox_tmp.copyTo(new_bbox);

    if (old_bbox.rows != new_bbox.rows || old_bbox.cols != new_bbox.cols) {
      continue;
    } else {
      double mse = 0;
      for (int i = 0; i < old_bbox.cols; i++) {
        for (int j = 0; j < old_bbox.rows; j++) {
          uchar a = old_bbox.at<uchar>(j, i);
          uchar b = new_bbox.at<uchar>(j, i);
          double diff = a - b;
          double rr = diff * diff;
          mse += rr;
        }
      }
      mse = mse / (old_bbox.cols * old_bbox.rows);
      if (mse > 250.0) {
        new_bgs_tracked_res.push_back(box);
      }
    }
  }
  imgGrayPrev = imgGray.clone();
  return new_bgs_tracked_res;
}

#endif // _MCDWRAPPER_CPP_